raft.go

```diff
@@ -1680,6 +1730,82 @@ func (r *Raft) requestVote(rpc RPC, req *RequestVoteRequest) {
 	r.setLastContact()
 }
 
+// requestPreVote is invoked when we get a request Pre-Vote RPC call.
+func (r *Raft) requestPreVote(rpc RPC, req *RequestPreVoteRequest) {
+	defer metrics.MeasureSince([]string{"raft", "rpc", "requestVote"}, time.Now())
+	r.observe(*req)
+
+	// Setup a response
+	resp := &RequestPreVoteResponse{
+		RPCHeader: r.getRPCHeader(),
+		Term:      r.getCurrentTerm(),
+		Granted:   false,
+	}
+	var rpcErr error
+	defer func() {
+		rpc.Respond(resp, rpcErr)
+	}()
+
+	// Check if we have an existing leader [who's not the candidate] and also
+	var candidate ServerAddress
+	candidateID := ServerID(req.ID)
+
+	// if the Servers list is empty that mean the cluster is very likely trying to bootstrap,
+	// Grant the vote
+	if len(r.configurations.latest.Servers) > 0 && !inConfiguration(r.configurations.latest, candidateID) {
+		r.logger.Warn("rejecting pre-vote request since node is not in configuration",
+			"from", candidate)
+		return
+	}
+
+	if leaderAddr, leaderID := r.LeaderWithID(); leaderAddr != "" && leaderAddr != candidate {
+		r.logger.Warn("rejecting pre-vote request since we have a leader",
+			"from", candidate,
+			"leader", leaderAddr,
+			"leader-id", string(leaderID))
+		return
+	}
+
+	// Ignore an older term
+	if req.Term < r.getCurrentTerm() {
+		return
+	}
+
+	if req.Term > r.getCurrentTerm() {
+		// continue processing here to possibly grant the pre-vote as in a "real" vote this will transition us to follower
+		r.logger.Debug("received a requestPreVote with a newer term, grant the pre-vote")
+		resp.Term = req.Term
+	}
+
+	// if we get a request for a pre-vote from a nonVoter  and the request term is higher, do not grant the Pre-Vote
+	// This could happen when a node, previously voter, is converted to non-voter
+	if len(r.configurations.latest.Servers) > 0 && !hasVote(r.configurations.latest, candidateID) {
+		r.logger.Warn("rejecting pre-vote request since node is not a voter", "from", candidate)
+		return
+	}
+
+	// Reject if their term is older
+	lastIdx, lastTerm := r.getLastEntry()
+	if lastTerm > req.LastLogTerm {
+		r.logger.Warn("rejecting pre-vote request since our last term is greater",
+			"candidate", candidate,
+			"last-term", lastTerm,
+			"last-candidate-term", req.LastLogTerm)
+		return
+	}
+
+	if lastTerm == req.LastLogTerm && lastIdx > req.LastLogIndex {
+		r.logger.Warn("rejecting pre-vote request since our last index is greater",
+			"candidate", candidate,
+			"last-index", lastIdx,
+			"last-candidate-index", req.LastLogIndex)
+		return
+	}
+
+	resp.Granted = true
+	r.setLastContact()
```

Comment:
Hmmm I didn't spot this before... but should we be setting this? I _think_ `lastContact` is supposed to indicate when we last heard from the _leader_ of the cluster. (Yep, I checked the doc comments for it) In a real vote, it seems reasonable to expect that the server we just gave our vote to is about to become the leader. In this case, that's not quite as clear (thought still likely I suppose).

Did you consider whether it's more like to be better or worse to set this? I think I'd lean towards not calling this because it changes state that at least some other part of raft relies on and so makes it just a tiny bit more risky than a pre-vote campaign could cause new behaviour in the cluster. I think semantically, the node that spoke to us is _not_ the leader right now either so it seems like we shouldn't treat it like it's already won an election. I guess that logic also applies to a real election though.

I don't think it is likely to be a problem either way. Just curious to think it through and see if we can imagine any impact or increased risk of impact either way. 🤔 

OK so it is used in a follower when heartbeat timeout is called when in follower state: https://github.com/hashicorp/raft/blob/cc2bb080d64f172ff2fcad13289e7ec07dd2e6c6/raft.go#L219

I think if we get here (granting the pre vote) then this node must already not have a leader (i.e. be in candidate state)... So I don't think it really is a _problem_, but I'm still not sure it makes sense to pretend we "heard from a leader" based on a prevote?

Reply:
yes I think you're right last contact should not be set in here, although I don't think of any scenario where it will create a breakage.

I will remove it to be on the safe side.